File: leetcode/java/numRookCaptures.java
class Solution {
    public int numRookCaptures(char[][] board) {
        int rookX = 0, rookY = 0, res = 0, i, j;
        for (i = 0; i < board.length; i++) {
            for (j = 0; j < board[i].length; j++) {
                if (board[i][j] == 'R') {
                    rookX = i;
                    rookY = j;
                }
            }
        }

        return find(board, rookX, rookY, res);
    }

    public int find(char[][] arr, int i, int j, int res) {
        int top = i, bottom = i, left = j, right = j;

        while (top >= 0) {
            if (arr[top][j] == 'B') break;
            else if (arr[top][j] == 'p') {
                res++;
                break;
            }
            top--;
        }

        while (bottom < arr.length) {
            if (arr[bottom][j] == 'B') break;
            else if (arr[bottom][j] == 'p') {
                res++;
                break;
            }
            bottom++;
        }

        while (left >= 0) {
            if (arr[i][left] == 'B') break;
            else if (arr[i][left] == 'p') {
                res++;
                break;
            }
            left--;
        }

        while (right < arr.length) {
            if (arr[i][right] == 'B') break;
            else if (arr[i][right] == 'p') {
                res++;
                break;
            }
            right++;
        }

        return res;
    }
}
//public static void main(String[] args) {
//        char[][] chars1 = {
//        {'.','.','.','.','.','.','.','.'},
//        {'.','.','.','p','.','.','.','.'},
//        {'.','.','.','R','.','.','.','p'},
//        {'.','.','.','.','.','.','.','.'},
//        {'.','.','.','.','.','.','.','.'},
//        {'.','.','.','p','.','.','.','.'},
//        {'.','.','.','.','.','.','.','.'},
//        {'.','.','.','.','.','.','.','.'}
//        };
//        char[][] chars2 = {
//        {'.','.','.','.','.','.','.','.'},
//        {'.','p','p','p','p','p','.','.'},
//        {'.','p','p','B','p','p','.','.'},
//        {'.','p','B','R','B','p','.','.'},
//        {'.','p','p','B','p','p','.','.'},
//        {'.','p','p','p','p','p','.','.'},
//        {'.','.','.','.','.','.','.','.'},
//        {'.','.','.','.','.','.','.','.'}
//        };
//        char[][] chars3 = {
//        {'.','.','.','.','.','.','.','.'},
//        {'.','.','.','p','.','.','.','.'},
//        {'.','.','.','p','.','.','.','.'},
//        {'p','p','.','R','.','p','B','.'},
//        {'.','.','.','.','.','.','.','.'},
//        {'.','.','.','B','.','.','.','.'},
//        {'.','.','.','p','.','.','.','.'},
//        {'.','.','.','.','.','.','.','.'}
//        };
//
//        System.out.println(new Solution().numRookCaptures(chars1));;
//        System.out.println(new Solution().numRookCaptures(chars2));;
//        System.out.println(new Solution().numRookCaptures(chars3));;
//        };

File: leetcode/ts/numRookCaptures.ts
function numRookCaptures(board: string[][]): number {
  let res = 0,
    rookX = 0,
    rookY = 0;

  for (let i = 0; i < board.length; i++) {
    for (let j = 0; j < board[i].length; j++) {
      if (board[i][j] === "R") {
        rookX = i;
        rookY = j;
        break;
      }
    }
  }

  return find(board, rookX, rookY, res);
}

function find(arr: string[][], i: number, j: number, res: number): number {
  if (typeof arr[i][j - 1] !== "undefined") {
    let left = j;
    while (left >= 0) {
      if (arr[i][left - 1] === "B") break;
      else if (arr[i][left - 1] === "p") {
        res++;
        break;
      }
      left--;
    }
  }

  if (typeof arr[i][j + 1] !== "undefined") {
    let right = j;
    while (right < arr[i].length) {
      if (arr[i][right + 1] === "B") break;
      else if (arr[i][right + 1] === "p") {
        res++;
        break;
      }
      right++;
    }
  }

  if (typeof arr[i - 1] && arr[i - 1][j] !== "undefined") {
    let top = i;
    while (top >= 0 && arr[top - 1]) {
      if (arr[top - 1][j] === "B") break;
      else if (arr[top - 1][j] === "p") {
        res++;
        break;
      }
      top--;
    }
  }

  if (typeof arr[i + 1] && arr[i + 1][j] !== "undefined") {
    let bottom = i;
    while (bottom < arr[i].length && arr[bottom + 1]) {
      if (arr[bottom + 1][j] === "B") break;
      else if (arr[bottom + 1][j] === "p") {
        res++;
        break;
      }
      bottom++;
    }
  }
  return res;
}

console.log(
  numRookCaptures([
    [".", ".", ".", ".", ".", ".", ".", "."],
    [".", ".", ".", "p", ".", ".", ".", "."],
    [".", ".", ".", "R", ".", ".", ".", "p"],
    [".", ".", ".", ".", ".", ".", ".", "."],
    [".", ".", ".", ".", ".", ".", ".", "."],
    [".", ".", ".", "p", ".", ".", ".", "."],
    [".", ".", ".", ".", ".", ".", ".", "."],
    [".", ".", ".", ".", ".", ".", ".", "."],
  ])
);

console.log(
  numRookCaptures([
    [".", ".", ".", ".", ".", ".", ".", "."],
    [".", "p", "p", "p", "p", "p", ".", "."],
    [".", "p", "p", "B", "p", "p", ".", "."],
    [".", "p", "B", "R", "B", "p", ".", "."],
    [".", "p", "p", "B", "p", "p", ".", "."],
    [".", "p", "p", "p", "p", "p", ".", "."],
    [".", ".", ".", ".", ".", ".", ".", "."],
    [".", ".", ".", ".", ".", ".", ".", "."],
  ])
);

console.log(
  numRookCaptures([
    [".", ".", ".", ".", ".", ".", ".", "."],
    [".", ".", ".", "p", ".", ".", ".", "."],
    [".", ".", ".", "p", ".", ".", ".", "."],
    ["p", "p", ".", "R", ".", "p", "B", "."],
    [".", ".", ".", ".", ".", ".", ".", "."],
    [".", ".", ".", "B", ".", ".", ".", "."],
    [".", ".", ".", "p", ".", ".", ".", "."],
    [".", ".", ".", ".", ".", ".", ".", "."],
  ])
);

console.log(
  numRookCaptures([
    [".", ".", ".", ".", ".", ".", ".", "."],
    [".", ".", ".", ".", ".", ".", ".", "."],
    [".", ".", ".", ".", ".", ".", ".", "."],
    [".", ".", ".", "R", ".", ".", ".", "."],
    [".", ".", ".", ".", ".", ".", ".", "."],
    [".", ".", ".", ".", ".", ".", ".", "."],
    [".", ".", ".", ".", ".", ".", ".", "."],
    [".", ".", ".", ".", ".", ".", ".", "."],
  ])
);
